File: all/349_intersection_of_two_arrays/sol_sorting_two_pointers.py
from bisect import bisect_left
from typing import List


class Solution:
    def intersection(self, nums1: List[int], nums2: List[int]) -> List[int]:
        answer_list = list()

        # Sort the nums1 list in the ascending order.
        nums1.sort()

        # Sort the nums2 list in the ascending order.
        nums2.sort()

        i = 0
        j = 0

        while (i < len(nums1)) and (j < len(nums2)):
            if (nums1[i]) == (nums2[j]):
                # If current iterated two values are the same, we might find a
                # new intersected value.
                if (not answer_list) or ((answer_list[-1]) != (nums1[i])):
                    # Add current iterated value to answer list if it is the
                    # first found intersected element or it is different from
                    # the previous found intersected element.
                    answer_list.append(nums1[i])

                # Increase both pointers by 1.
                i += 1
                j += 1
            elif (nums1[i]) < (nums2[j]):
                # If current iterated value in the nums1 list is less than
                # current iterated value in the nums2 list, we need to increase
                # the pointer for nums1 list by 1 since that value will not show
                # in the rest of sections in the nums2 list.
                i += 1
            else:
                # If current iterated value in the nums2 list is greater than
                # current iterated value in the nums2 list, we need to increase
                # the pointer for nums2 list by 1 since that value will not
                # show in the rest of sections in the nums1 list.
                j += 1

        return answer_list


if '__main__' == __name__:
    sol = Solution()
    testcases_list = [
        [[1, 2, 2, 1], [2, 2]],
        [[4, 9, 5], [9, 4, 9, 8, 4]],
    ]

    for testcase in testcases_list:
        print(sol.intersection(*testcase))

File: all/349_intersection_of_two_arrays/sol_sorting_two_pointers.c
#include <stdbool.h>
#include <stddef.h>
#include <stdlib.h>

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
typedef struct AnswerArrayStruct {
    int *array_ptr;
    int array_length;
    int next_available_idx;
} AnswerArrayST;

int compareFunction(const void *p1, const void *p2) {
    const int i1 = (*((const int*)p1));
    const int i2 = (*((const int*)p2));

    return ((i1 > i2) - (i1 < i2));
}

bool addElementToAnswerArrayST(
    AnswerArrayST * const answer_array_st,
    const int x
) {
    int *tmp_array_ptr = NULL;
    int new_array_length = 0;

    if (
        (answer_array_st->next_available_idx)
        >=
        (answer_array_st->array_length)
    ) {
        if (0 == (answer_array_st->array_length)) {
            new_array_length = 4;
        }
        else {
            new_array_length = ((answer_array_st->array_length) << 1);
        }

        tmp_array_ptr = (
            (int*)realloc(
                answer_array_st->array_ptr,
                sizeof(*(answer_array_st->array_ptr)) * new_array_length
            )
        );

        if (NULL == tmp_array_ptr) {
            return false;
        }

        answer_array_st->array_length = new_array_length;
        answer_array_st->array_ptr = tmp_array_ptr;
    }

    (answer_array_st->array_ptr)[(answer_array_st->next_available_idx)++] = x;

    return true;
}

int* intersection(
    int* nums1,
    int nums1Size,
    int* nums2,
    int nums2Size,
    int* returnSize
) {
    AnswerArrayST answer_array_st = {
        .array_ptr = NULL,
        .array_length = 0,
        .next_available_idx = 0,
    };
    int previous_same_element = 0;

    /* Sort the nums1 array in the ascending order. */
    qsort((void*)nums1, (size_t)nums1Size, sizeof(int), &compareFunction);

    /* Sort the nums2 array in the ascending order. */
    qsort((void*)nums2, (size_t)nums2Size, sizeof(int), &compareFunction);

    for (
        int idx_1 = 0, idx_2 = 0;
        (idx_1 < nums1Size) && (idx_2 < nums2Size);
    ) {
        if (nums1[idx_1] == nums2[idx_2]) {
            /*
             * If current iterated two values are the same, we might find a new
             * intersected value.
             */
            if (
                (0 == answer_array_st.next_available_idx)
                ||
                (previous_same_element != nums1[idx_1])
            ) {
                /*
                 * Add current iterated value to answer array if it is the first
                 * found intersected element or it is different from the
                 * previous found intersected element.
                 */
                addElementToAnswerArrayST(&answer_array_st, nums1[idx_1]);

                /* Update previous found intersected element. */
                previous_same_element = nums1[idx_1];
            }

            /* Increase both pointers by 1. */
            ++idx_1;
            ++idx_2;
        }
        else if (nums1[idx_1] < nums2[idx_2]) {
            /*
             * If current iterated value in the nums1 array is less than current
             * iterated value in the nums2 array, we need to increase the
             * pointer for nums1 array by 1 since that value will not show in
             * the rest of sections in the nums2 array.
             */
            ++idx_1;
        }
        else {
            /*
             * If current iterated value in the nums2 array is greater than
             * current iterated value in the nums2 array, we need to increase
             * the pointer for nums2 array by 1 since that value will not show
             * in the rest of sections in the nums1 array.
             */
            ++idx_2;
        }
    }

    (*returnSize) = (answer_array_st.next_available_idx);

    return (answer_array_st.array_ptr);
}
